### ISSUE TITLE
fix: Opening in new tab not working reliably in Safari for iOS

### ISSUE THREAD

**Issue by inf0rmatix** [Bdaya-Dev/oidc#31]
**Description**

When using the package in a web app the new tab does not open reliably. Sometimes it doesn't open at all, sometimes I have to make the call for `manager.loginAuthorizationCodeFlow()` multiple times until it works once.

The issue here is safari's popup blocking behavior. You must not have any asynchronous code between the user tap event and launching the url. Otherwise it will be most likely blocked by safari. Causing this problem is awaiting the calls to `store.setStateData` and `store.setCurrentNonce` which, with the default OidcStore, uses the local storage of the browser. That is reason enough for safari to block the new tab which is being tried to open AFTER writing to the store.

**Steps To Reproduce**

1. Setup Oidc the default way in a web app.
2. Use `loginAuthorizationCodeFlow` with the default parameters (opening in a new tab)
3. Open the app in Safari on iOS
4. Try to login. 
5. Inconsistent behavior of the new tab opening or being blocked.

**Expected Behavior**
The new tab should always open.

**Comment by ahmednfwela** [comment]
that is an interesting problem, I guess the popup policies are stricter on phone than on pc.
and I guess you can always use `samePage` navigation mode to circumvent this.

there is a problem with not waiting for `store.setStateData` and `store.setCurrentNonce` though;
since 
1. there may be an error when saving the state/nonce, and the new tab will get launched either way, causing an invalid state even if the auth is successful.
2. there might be a race condition especially in `samePage` navigation mode, where the app will get destroyed when redirecting before it's able to save the state/nonce.

**Comment by inf0rmatix** [comment]
Yes, Samepage works, of course. 

Yes, storing the data might fail silently now.

Ideally, the whole logic would be refactored so that:

The nonce and key-pair are created and saved when initiating, thus making the loginCall Synchronous until calling `launchUrl`.

**Comment by ahmednfwela** [comment]
do you have any sources on why safari blocks new tabs ? maybe I can think of something

**Comment by inf0rmatix** [comment]
No, I don’t have any links to valid first hand Apple information.
I have read numerous other Issues and stack overflow posts.

The main Issue seems to be:
Doing any „await“ stuff before opening the new tab

Am 24.10.2023 um 09:15 schrieb Ahmed Fwela ***@***.***>:



do you have any sources on why safari blocks new tabs ? maybe I can think of something

**Comment by jlambright** [comment]
There is a similar issue with Firefox as well, which blocks the execution of JavaScript to close windows that were NOT opened by the script itself. This causes behavior where the new login/out tabs don't close. I did see some warning in the Chrome console that this behavior may be coming to Chromium based browsers in the near future.

**Comment by ahmednfwela** [comment]
@jlambright to be clear this issue only happens with new tab and popup navigation, and not with same tab navigation, right?

**Comment by jlambright** [comment]
Yes. Though we do see it open a background tab in Firefox (for logout only)
even on same page navigation.

**Comment by ahmednfwela** [comment]
I have opened a separate issue to track this here https://github.com/Bdaya-Dev/oidc/issues/68 
since this issue is mentioning a problem specific to Safari for iOS not opening a new tab at all
